### PR TITLE
Friendship v2

### PR DESCRIPTION
**Pull request by jonathastavares**
This is the 3rd milestone for Ruby on Rails Social Media project.

In this milestone I added:

- New associations methods for user model to use Database storage instead of Ruby logic.

This project was built using:

- Ruby
- Ruby on Rails
- Devise
- RSpec
- Capybara

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -12,22 +12,10 @@ class User < ApplicationRecord
...
-    friendships.map { |friendship| friendship.friend unless friendship.status }.compact
-  end
-
-  # Users who have requested to be friends
-  def friend_requests
-    inverse_friendships.map { |friendship| friendship.user unless friendship.status }.compact
-  end
+  has_many :confirmed_friendships, -> { where status: true }, class_name: 'Friendship'
+  has_many :friends, through: :confirmed_friendships
+  has_many :pending_friends, -> { where status: nil }, class_name: 'Friendship'
```
Comment (GzimAsani) [inline app/models/user.rb]: Add in this line also the `foreign_key: "friend_id"`
example:
`has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"`

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# Status: Changes Required :red_circle: 

## Hello @jonathastavares, your project looks great you've done awesome job with your models, however I have a list of requirements for you which will help you:

* [x] When friendship is confirmed, an additional row with reversed user-friend pair is created

```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```
* [x] thanks to this, we can find user friends only by user_id in the friendships table <- **You have already implemented this!** I just thought to mention and show you the way

```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```
* [x] in order to get all posts written by user on their friends, you can use the simple method in the User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

### If you have any questions, feel free to comment  here and I will try to respond ASAP

# HAPPY CODING :computer: :desktop_computer:

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, thank you for working on the changes suggested previously :ok_hand:
## To Highlight 🎉
- [ ] Git flow is used properly.
- [ ] README includes every instruction to run the project locally.
## To Improve 🙏
- [ ] When friendship is confirmed, an additional row with reversed user-friend pair is created in database. Currently, there is no mutual friendship implemented. Kindly fix that
# Optional
- [ ] You may distribute spec tests in separate files. So User model specs are in its own file and so on.
### Happy Coding ☺️
## If you have any questions you can comment here in this PR or can always connect with me on slack ( @Sajjad Ahmad ).

**Review by george-shammar — APPROVED** [review]
### **STATUS: APPROVED** :heavy_check_mark: :heavy_check_mark: 

@jonathastavares , this is just awesome!!!!

You've ticked all the boxes by:
:heavy_check_mark: Making your repo look professional.
:heavy_check_mark: Implementing the required changes. Friendship can now be confirmed.
:heavy_check_mark: I like the fact that you included integration testing.

Good job. :+1: 
I'm happy to approve this milestone.

You may merge this branch and proceed to the next milestone.

Cheers :smile With respect :tophat:

**Pull request merged**